builtin/provisioners/chef-client/resource_provisioner.go

```diff
@@ -0,0 +1,402 @@
+		p.runChefClient = p.runChefClientFunc(windowsConfDir)
+	default:
+		return fmt.Errorf("Unsupported connection type: %s", s.Ephemeral.ConnInfo["type"])
+	}
+
+	// Get a new communicator
+	comm, err := communicator.New(s)
+	if err != nil {
+		return err
+	}
+
+	// Wait and retry until we establish the connection
+	err = retryFunc(comm.Timeout(), func() error {
+		err := comm.Connect(o)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	defer comm.Disconnect()
+
+	if !p.SkipInstall {
+		if err := p.installChefClient(o, comm); err != nil {
+			return err
+		}
+	}
+
+	o.Output("Creating configuration files...")
+	if err := p.createConfigFiles(o, comm); err != nil {
+		return err
+	}
+
+	o.Output("Starting initial Chef-Client run...")
+	if err := p.runChefClient(o, comm); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Validate checks if the required arguments are configured
+func (r *ResourceProvisioner) Validate(c *terraform.ResourceConfig) (ws []string, es []error) {
+	p, err := r.decodeConfig(c)
+	if err != nil {
+		es = append(es, err)
+		return ws, es
+	}
+
+	if p.NodeName == "" {
+		es = append(es, fmt.Errorf("Key not found: node_name"))
+	}
+	if p.RunList == nil {
+		es = append(es, fmt.Errorf("Key not found: run_list"))
+	}
+	if p.ServerURL == "" {
+		es = append(es, fmt.Errorf("Key not found: server_url"))
+	}
+	if p.ValidationClientName == "" {
+		es = append(es, fmt.Errorf("Key not found: validation_client_name"))
+	}
+	if p.ValidationKeyPath == "" {
+		es = append(es, fmt.Errorf("Key not found: validation_key_path"))
+	}
+
+	return ws, es
+}
+
+func (r *ResourceProvisioner) decodeConfig(c *terraform.ResourceConfig) (*Provisioner, error) {
+	p := new(Provisioner)
+
+	decConf := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           p,
+	}
+	dec, err := mapstructure.NewDecoder(decConf)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := dec.Decode(c.Raw); err != nil {
+		return nil, err
+	}
+
+	if p.Environment == "" {
+		p.Environment = "_default"
+	}
+
+	if attrs, ok := c.Raw["attributes"]; ok {
+		p.Attributes, err = rawToJSON(attrs)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing the attributes: %v", err)
+		}
+	}
+
+	return p, nil
+}
+
+func rawToJSON(raw interface{}) (interface{}, error) {
+	switch s := raw.(type) {
+	case []map[string]interface{}:
+		if len(s) != 1 {
+			return nil, errors.New("unexpected input while parsing raw config to JSON")
+		}
+
+		var err error
+		for k, v := range s[0] {
+			s[0][k], err = rawToJSON(v)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return s[0], nil
+	default:
+		return raw, nil
+	}
+}
+
+// retryFunc is used to retry a function for a given duration
+func retryFunc(timeout time.Duration, f func() error) error {
+	finish := time.After(timeout)
+	for {
+		err := f()
+		if err == nil {
+			return nil
+		}
+		log.Printf("Retryable error: %v", err)
+
+		select {
+		case <-finish:
+			return err
+		case <-time.After(3 * time.Second):
```

Comment:
So this retries every 3 seconds until it reaches the given timeout value? 3s seems arbitrary - how do you feel about an exponential backoff instead?

So, instead of providing a hard timeout (like 30s), you provide the number of backoffs (like 4), and each backoff is the square of the previous, starting at something like 3s, so:

```go
func retryFunc(times int, stopCh chan struct{}, f func() error) error {
  backoff := 3

  for i := o; i < times; i++ {
    err := f()
    if err == nil {
      return nil
    }
    log.Printf("retrying...")

    select {
      case <-stopCh:
        return err
      case <-time.After(backoff * time.Second):
        backoff = backoff * backoff
        continue
    }
  }
}
```

Even if you don't like this approach, I think we still need some kind of termination channel in here, or else this will block if the user cancels.